connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java

```diff
@@ -530,13 +530,33 @@ private void pauseAll() {
         consumer.pause(consumer.assignment());
     }
 
+    private void put(Collection<SinkRecord> records) {
+        retryWithToleranceOperator.reset();
+        retryWithToleranceOperator.execute(() -> {
+            try {
+                task.put(records);
```

Comment:
we cannot retry with the same batch of records. this is error prone, especially for connectors that provide exactly once guarantees (it may end up writing up multiple copies of the same record into the same destination, IIUC). also, with this approach we are not sure which record (if any) is the bad one. in the ideal world, a bad record must be identified and written to the DLQ or retried appropriately. I think a better approach would be to provide APIs to the sink tasks that allow it to report bad records to the DLQ, and delegate retries to the task implementation, as the retry logic depends a lot on the task implementation. there are multiple things that can go wrong in the put(), and retrying them all is a bit dangerous. also, retriable exceptions are already retried by the framework, (look at ) so I'm not sure what is buying us. 

If the intent is to simply report errors, then we should come up with a way to directly report errors. 

Either way, these changes would need a KIP and subsequent discussions to determine the way to go forward.